### ISSUE TITLE
Meta: CircuitInputBuilder progress status

### ISSUE THREAD

**Issue by ed255** [privacy-scaling-explorations/zkevm-circuits#222]
The CircuitInputBuilder is the module responsible for generating all the circuit inputs required to generate a proof.  Since we have different circuits, the CircuitInputBuilder should be able to generate the inputs for all of them.

## Circuits and inputs
- [EVM Circuit](https://github.com/han0110/zkevm-circuits/blob/3c57d83839d68064d81fea52ef9a55ec82a3ef5b/zkevm-circuits/src/evm_circuit/execution.rs#L68) :
    - Inputs: Block -> Transactions -> Calls, ExecutionSteps
- [State Circuit](https://github.com/appliedzkp/zkevm-circuits/blob/1ef27da64d823c868835a90f7f65d8e4ebd5786f/zkevm-circuits/src/state_circuit/state.rs?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L74):
    - Inputs: Operation Container: MemoryOps, StackOps, StorageOps, etc.
- MPT Circuit:
    - Inputs: MPT Proofs for a subset of the StorageOps and AccountOps

## Steps

1. Query geth for Block, Txs and TxExecTraces
    - [x] RPC Calls for `eth_getBlockByNumber`, `debug_traceBlockByNumber`
2. Get State Accesses from TxExecTraces
    - [x] `gen_state_access_trace`, `AccessSet`
3. Query geth for all accounts, storage keys, and codes from Accesses
    - [x] RPC Calls for `eth_getProof`, `eth_getCode`
4. Build a partial StateDB from 3
    - [x] Fill `StateDB` from `AccessSet`
5. For each step in TxExecTraces, gen the associated ops and state circuit inputs
    - [ ] generate a list of hashing sources (tx, code source, keccak opcode) that the keccack circuit must verify https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/579
    - [x] https://github.com/appliedzkp/zkevm-circuits/issues/207#L551
    - [x] https://github.com/appliedzkp/zkevm-circuits/issues/174 (blocked by "implement a return with failure opcode in bus-mapping")
    - [ ] Generate list of extra account read accesses (from failed CREATE* due to address collision)
    - [ ] https://github.com/appliedzkp/zkevm-circuits/issues/477
        - [x] `POP`, `MSTORE8`
        - [x] `MSIZE`, `Virt BeginTx`
        - rest of opcodes: implementation of the `gen_associated_ops` for each
          opcode should be done by the same person that implements such opcode
          in the circuit, and ideally submitted in the same PR.
            - [x] CALL
            - [x] STOP [spec PR](https://github.com/appliedzkp/zkevm-specs/pull/100)
    - Ops for
        - [x] Memory
        - [x] Stack
        - [x] Storage
        - [x] TxAccessListAccount
            - On failure/revert: need to generate the reverse op
            - Content: write, TxId, address, prev: bool, current: bool
            - Only in:
                - BeginTx
                - `EXTCODECOPY` (address on stack[-1])
                - `EXTCODESIZE` (address on stack[-1])
                - `BALANCE` (address on stack[-1])
                - `SELFDESTRUCT` (address on stack[-1])
                - `*CALL*` (address on stack[-2])
                - `CREATE*` (address we need to calculated ourself in case collision)
        - [x] TxAccessListStorageSlot https://github.com/appliedzkp/zkevm-circuits/issues/173
            - Content: write, TxId, address, key, prev: bool, current: bool
            - Only in:
                - SSTORE
                - SLOAD
        - [x] #1008
            - Content: read/write, TxId, prev: word, current: word
            - NOTE: Trace doesn't provide the refund value, so we need to
              calculate it based on the different situations.
            - NOTE: Defer the implementation of this Op in the bus-mapping to
              the same person that implements the TxRefund logic in the
              circuits.
            - NOTE: It only writes when `is_persistent`, no need to undo when failure/revert
            - Only in:
                - write (rules in [EIP3529](https://eips.ethereum.org/EIPS/eip-3529))
                    - SSTORE with value = 0
                - read (when end of tx, do refund to tx sender)
                    - STOP
                    - RETURN
                    - REVERT
        - [x] Account
            - Content: read/write, address, tag (nonce, balance, code_hash),
              prev: word, curr: word
            - Only in:
                - BeginTx
                - `EXTCODECOPY` (address on stack[-1])
                - `EXTCODESIZE` (address on stack[-1])
                - `BALANCE` (address on stack[-1])
                - `SELFDESTRUCT` (address on stack[-1])
                - `*CALL*` (address on stack[-2])
                - `CREATE*` (address we need to calculated ourself in case collision)
                - In end of transaction, TxRefund to tx.from, and miner reward
                  (balance update):
                    - `STOP`
                    - `RETURN`
                    - `REVERT`
        - [x] AccountDestructed
            - Content: write, TxId, address, prev: bool, current: bool
            - NOTE: It only writes when `is_persistent`, no need to undo when
              failure/revert
            - Only in:
                - `SELFDESTRUCT`
        - [x] CallContext (https://github.com/appliedzkp/zkevm-circuits/issues/225)
6. Query geth for extra accounts from 5
    - [x] RPC Calls for `eth_getProof`
7. Merge results from 3 and 6 into a Partial State Trie to generate proofs for the circuit.
    - [ ] Extend `AccessSet` from 4 with results from 5
        - Collect all proofs from `AccessSet`, reusing results from 3
    - [x] Implement a Partial State Trie that supports updates and proof generation https://github.com/miha-stopar/mpt
    - [ ] Build a Partial State Trie from collected proofs
    - [ ] Generate intermediate proofs for StorageOps and AccountOps (Blocked
      by inputs / witness spec of MPT Circuit
      https://github.com/appliedzkp/zkevm-circuits/tree/mpt2)
          - Sort all StorageOps, AccountOps, AccountDeletedOps by (address, key, rwc)
          - Use wrapper over MPT witness generation https://github.com/miha-stopar/mpt
          - For each unique (address, key), get a "before" proof, apply last write, get "after" proof
          - Sort all AccountOps by (address, rwc)
          - For each unique (address), get "before" proof, apply write with final values, get "after" proof
          - Append all hashes used by the MPT into the hash source list.
8. Transform output from `CircuitInputBulder` into the types that the circuit use
    - [x] Write functions to transform types.  https://github.com/appliedzkp/zkevm-circuits/pull/239

### Circuit Steps VS Geth Exec Steps

Currently the circuits don't use a 1 to 1 match between geth exec steps and
circuit steps.  A circuit step is known as [`ExecutionState` in the circuit
spec](https://github.com/appliedzkp/zkevm-specs/blob/master/src/zkevm_specs/evm/execution_state.py#L7).
The steps used in the circuit are any of the following kind:
- **Success step by opcode**:  Match 1 to 1 with geth exec steps when there's
  no error, alghouth some opcodes are merged into the same one, like {ADD,
  SUB}, or {LT, GT, EQ}.
- **Success Multi-Step by opcode**: One successful opcode execution that uses a
  dynamic array as argument gets transformed into a list of smaller multi-step
  opcodes that perform the operation on the dynamic array by chunks of fixed
  size.  The affected opcodes are:
    - CODECOPY
    - EXTCODECOPY
    - CALLDATACOPY
    - RETURNDATACOPY
    - CREATE
    - CREATE2
    - SHA3
    - RETURN
    - REVERT
- **Implicit state transitions**: There are some state transitions that happen
  implicity without an opcode executon.  They are:
    - BeginTx [spec
      impl](https://github.com/han0110/zkevm-specs/blob/feature%2Fcall-and-stop/src/zkevm_specs/evm/execution/begin_tx.py).
- **Errors caused by opcodes**: When an opcode execution causes an error, the
  step contains the witness to prove that the error happened:

Tests of some of these `ExecutionState` can be [found here](https://github.com/han0110/zkevm-specs/tree/feature/call-and-stop/tests/evm)

Tasks:
- [x] Generalize the output of the CircuitInputBuilder to follow the definition of `ExecutionState`. (issue: https://github.com/appliedzkp/zkevm-circuits/issues/309, PR: https://github.com/appliedzkp/zkevm-circuits/pull/393)
    - For example, this could be done by converting the ExecStep into an enum
      that can be: ExecStep, MultiStep, ErrorStep, VirtualStep.

**Comment by ed255** [comment]
> @ed255 Do you think we can close this?

Yes! The only missing part of this issue is point 7, which relates to the integration of the MPT, so I've moved that to the issue https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/811#issuecomment-1531556675 and we can now close this :)
